File: Data_extraction/extraction_scripts/mongodb/src/main/java/org/tonde/Catalogue.java
package org.tonde;


import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.CreateCollectionOptions;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.ValidationOptions;
import org.bson.BsonType;
import org.bson.Document;
import org.bson.conversions.Bson;

import java.util.ArrayList;

public class Catalogue {

    public static void createCollCatalogue(MongoDatabase db) {

        Bson marqueType = Filters.type("Marque", BsonType.STRING);
        Bson nomType = Filters.type("Nom", BsonType.STRING);
        Bson puissanceType = Filters.type("Puissance", BsonType.DOUBLE);
        Bson longueurType = Filters.type("Longueur", BsonType.STRING);
        Bson nbPlacesType = Filters.type("NbPlaces", BsonType.INT32);
        Bson nbPortesType = Filters.type("NbPortes", BsonType.INT32);
        Bson Couleurtype = Filters.type("Couleur", BsonType.STRING);
        Bson Occasiontype = Filters.type("Occasion", BsonType.BOOLEAN);
        Bson Prixtype = Filters.type("Prix", BsonType.DOUBLE);


        Bson validator = Filters.and(marqueType, nomType, puissanceType,longueurType,nbPlacesType,nbPortesType,Couleurtype,Occasiontype,Prixtype);

        db.getCollection("Catalogue").drop();
        db.createCollection("Catalogue", new CreateCollectionOptions().validationOptions(new ValidationOptions()
                .validator(validator)));
    }

    public static void insertCatalogue (ArrayList<String[]> catalogueArr, MongoCollection<Document> collection) {
        for (String[] vehicule : catalogueArr) {
            Document doc = new Document()
                    .append("Marque",vehicule[0])
                    .append("Nom",vehicule[1])
                    .append("Puissance",Float.parseFloat(vehicule[2]))
                    .append("Longueur",vehicule[3])
                    .append("NbPlaces",Integer.parseInt(vehicule[4]))
                    .append("NbPortes",Integer.parseInt(vehicule[5]))
                    .append("Couleur",vehicule[6])
                    .append("Occasion",Boolean.parseBoolean(vehicule[7]))
                    .append("Prix",Float.parseFloat(vehicule[8]));

            collection.insertOne(doc);
        }
    }
}

File: BigData/extraction_scripts/insertion_catalogue_immatriculation/catalogue.py
from pymongo import MongoClient
from pymongo.collection import Collection
from pymongo.database import Database
from bson import Type
import csv

class Catalogue:
    """Classe pour insérer les données du catalogue dans la base de données MongoDB"""

    @staticmethod
    def create_coll_catalogue(db: Database):
        """Crée la collection Catalogue avec un schéma de validation"""
        validator = {
            "$jsonSchema": {
                "bsonType": "object",
                "required": ["Marque", "Nom", "Puissance", "Longueur", "NbPlaces", "NbPortes", "Couleur", "Occasion", "Prix"],
                "properties": {
                    "Marque": {"bsonType": "string"},
                    "Nom": {"bsonType": "string"},
                    "Puissance": {"bsonType": "double"},
                    "Longueur": {"bsonType": "string"},
                    "NbPlaces": {"bsonType": "int"},
                    "NbPortes": {"bsonType": "int"},
                    "Couleur": {"bsonType": "string"},
                    "Occasion": {"bsonType": "bool"},
                    "Prix": {"bsonType": "double"}
                }
            }
        }

        db.drop_collection("Catalogue")
        db.create_collection("Catalogue", validator=validator)

    @staticmethod
    def insert_catalogue(catalogue_arr, collection: Collection):
        """Insère les données du catalogue dans la collection Catalogue"""

        for vehicule in catalogue_arr:
            doc = {
                "Marque": vehicule[0],
                "Nom": vehicule[1],
                "Puissance": float(vehicule[2]),
                "Longueur": vehicule[3],
                "NbPlaces": int(vehicule[4]),
                "NbPortes": int(vehicule[5]),
                "Couleur": vehicule[6],
                "Occasion": vehicule[7].lower() == 'true',
                "Prix": float(vehicule[8])
            }
            collection.insert_one(doc)

def load_catalogue_data_from_file(file_path):
    """Charge les données du catalogue depuis un fichier CSV"""
    catalogue_arr = []
    with open(file_path, mode='r') as file:
        reader = csv.reader(file)
        next(reader)  # Skip header row
        for row in reader:
            catalogue_arr.append(row)
    return catalogue_arr

if __name__ == "__main__":
    # client = MongoClient("mongodb://localhost:27017/")
    #docker 
    client = MongoClient("mongodb://mongo:27017/")
    db = client["bigdata_exam"]

    Catalogue.create_coll_catalogue(db)
    catalogue_data = load_catalogue_data_from_file("Catalogue.csv")
    collection = db["Catalogue"]
    Catalogue.insert_catalogue(catalogue_data, collection)
